nexus_constructor/main_window.py

```diff
@@ -240,6 +240,10 @@ def _update_3d_view_with_component_shapes(self):
             self.sceneWidget.add_transformation(component)
 
     def show_add_component_window(self, group: Group, new_group: bool):
+        if hasattr(self, "add_component_window") and not self.add_component_window.isHidden():
+            self.add_component_window._rejected()
+            self.add_component_window.setHidden(True)
+
```

Comment:
This is just a variation of the hack you were unhappy with last week.
It is also just papering over the cracks.

The question to ponder is: why are we creating a new AppComponentDialog every time?

Reply:
I see the point - the answer is because each component is different in terms of parent node, type etc.  You can see in :120 when we edit a group the difference is we pass a different flag which states whether we use empty groups or existing groups.  It doesn't simply 'fill in' the dialog/window etc, but creates a new object that has many members initialised with particular values which are important for various methods in the dialog (for example, remembering where the component goes after it is made, or remembering the initial state of the group so it can be restored if the user cancels it).  However, there are various object inside the AddComponentDialog class, such as the main UI layout, that mean it makes sense to have as a separate part (rather than as part of this MainWindow class).  Finally, the dialog is deleted afterwards so we are not getting a memory leak or anything like that.

I've removed the hasattr check, and now we have have a clean and consistent way of handling the dialog

Reply:
This is not clean.

Repeated clicks of the edit button each create a new AppComponentDialog which are possibly not destructed (I see the memory usage increase on my machine at least).

It should not be necessary to construct a new AppComponentDialog every time just to pass some values into it, i.e. add a method something like `def set_initial_values(group, is_new_group)`.